base/utf.jl

```diff
@@ -0,0 +1,917 @@
+    cnum = 0
+    for i = 1:len
+        @inbounds cnum += !is_surrogate_trail(d[i])
+    end
+    cnum
+end
+
+function endof(s::UTF16String)
+    d = s.data
+    i = length(d) - 1
+    i == 0 && return i
+    return is_surrogate_char(d[i]) ? i-1 : i
+end
+
+get_supplementary(lead::Unsigned, trail::Unsigned) = (UInt32(lead-0xd7f7)<<10 + trail)
+
+function next(s::UTF16String, i::Int)
+    ch = s.data[i]
+    !is_surrogate_char(ch) && return (Char(ch), i+1)
+    # check length, account for terminating \0
+    i >= (length(s.data)-1) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, i, UInt32(ch))
+    !is_surrogate_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, i, ch)
+    ct = s.data[i+1]
+    !is_surrogate_trail(ct) && utf_errfunc(UTF_ERR_NOT_TRAIL, i, ch)
+    Char(get_supplementary(ch, ct)), i+2
+end
+
+function reverseind(s::UTF16String, i::Integer)
+    j = length(s.data) - i
+    return is_surrogate_trail(s.data[j]) ? j-1 : j
+end
+
+lastidx(s::UTF16String) = length(s.data) - 1 # s.data includes NULL terminator
+
+function reverse(s::UTF16String)
+    d = s.data
+    out = similar(d)
+    out[end] = 0 # NULL termination
+    n = length(d)
+    @inbounds for i = 1:n-1
+        ch = d[n-i]
+        if is_surrogate_lead(ch)
+            out[i],out[i-1] = out[i-1],ch
+        else
+            out[i] = ch
+        end
+    end
+    UTF16String(out)
+end
+
+next(s::UTF32String, i::Int) = (s.data[i], i+1)
+endof(s::UTF32String) = length(s.data) - 1
+length(s::UTF32String) = length(s.data) - 1
+
+const UTF_NO_LONG_NULL = 1      # don't accept 0xc0 0x80 for '\0'
+const UTF_NO_SURROGATES = 2     # don't accept surrogate pairs in UTF-8/UTF-32
+const UTF_ACCEPT_LONG = 4       # accept long encodings (other than long null in UTF-8)
+
+const UTF_LONG = 1              # Long encodings are present
+const UTF_LATIN1 = 2            # characters in range 0x80-0xFF present
+const UTF_UNICODE2 = 4          # characters in range 0x100-0x7ff present
+const UTF_UNICODE3 = 8          # characters in range 0x800-0xd7ff, 0xe000-0xffff
+const UTF_UNICODE4 = 16         # non-BMP characters present
+const UTF_SURROGATE = 32        # surrogate pairs present
+
+# Get a UTF-8 continuation byte, give error if invalid, and update position and character value
+@inline function get_continuation(ch::UInt32, str, pos)
+    byt::UInt8 = str[pos += 1]
+    !is_valid_continuation(byt) && utf_errfunc(UTF_ERR_CONT, pos, byt)
+    (ch << 6) | (byt & 0x3f), pos
+end
+
+#=
+@doc """
+@brief      Validates and calculates number of characters in a string
+
+@param[in]  str     Vector of UInt8
+@param[in]  options flags to determine error handling (default 0)
+
+@return     (total characters, flags, 4-byte, 3-byte, 2-byte)
+@throws     ArgumentError
+""" ->
+=#
+function check_string_utf8(dat::Vector{UInt8}, options::Integer=0)
+    local byt::UInt8
+    local ch::UInt32, surr::UInt32
+    local totalchar::Int=0, num2byte::Int=0, num3byte::Int=0, num4byte::Int=0, flags::UInt=0
```

Comment:
OK, originally I had those setting to unsigned values, but ran into trouble because setting to `0x0` would make it think it was a `UInt8`, so I decorated everything with `UInt`, and then later changed to `Int`, because of issues with `sizeof(xxx)` returning an Int in the functions that called these functions.